src/cargo/core/workspace.rs

```diff
@@ -506,13 +528,43 @@ impl<'gctx> Workspace<'gctx> {
     /// Returns the root `[patch]` section of this workspace.
     ///
     /// This may be from a virtual crate or an actual crate.
-    pub fn root_patch(&self) -> CargoResult<HashMap<Url, Vec<Dependency>>> {
-        let from_manifest = match self.root_maybe() {
+    pub fn root_patch(&self) -> CargoResult<HashMap<Url, Vec<(Dependency, PathOrDefinition)>>> {
+        let patch_cv = self.gctx.get_cv_with_env(&ConfigKey::from_str("patch"))?;
+        let manifest_path = PathOrDefinition::Path(self.root_manifest().into());
+
+        let patch_from_manifest = match self.root_maybe() {
             MaybePackage::Package(p) => p.manifest().patch(),
             MaybePackage::Virtual(vm) => vm.patch(),
         };
 
-        let from_config = self.config_patch()?;
+        // Insert location information for each dep in the patch
+        let mut from_config: HashMap<_, Vec<_>> = HashMap::new();
+        for (url, dep) in self.config_patch()? {
+            from_config.insert(
+                url,
+                dep.iter()
+                    .map(|dep| {
+                        (
+                            dep.clone(),
+                            PathOrDefinition::Definition(
+                                patch_cv.as_ref().unwrap().definition().clone(),
+                            ),
+                        )
+                    })
+                    .collect(),
+            );
+        }
+
+        let mut from_manifest: HashMap<_, Vec<_>> = HashMap::new();
+        for (url, dep) in patch_from_manifest {
+            from_manifest.insert(
+                url.clone(),
+                dep.iter()
+                    .map(|dep| (dep.clone(), manifest_path.clone()))
+                    .collect(),
+            );
+        }
+
```

Comment:
This was
1. manifest
2. config
3. config
4. manifest

But became
1. config
2. manifest
5. manifest
6. config
7. manifest
8. config
9. manifest

Already the original code wasn't idea, it should either be completely parallel or consistently interleaved.  As the code grows, I would suggest interleaving makes the code harder to follow and we should be completely parallel.